---
title: "toetsRNAseq"
author: "Wouter&Bas"
date: "11-4-2022"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(Rsubread)
library(DESeq2)
library(here)
library(pheatmap)
library(org.Hs.eg.db)
library(GO.db)
library(GOstats)
library(png)
library(grid)
library(gridExtra)
```

### onderdeel 1 introductie
De onderzoeksvraag is: wat is het verschil in genexpressie in fibroblasten tussen behandeling met BCLXL en ONECUT3? Er zijn in de dataset bij twee type fibroblasten RNAseq uitgevoerd waarbij voor elke conditie een duplo meting is gedaan. de datasets zijn gemaakt van een stranded protocol en bestaan uit paired reads. 

### onderdeel 2 fastqc conclusie
```{r, include=FALSE, eval=FALSEfig.height=3, out.width= '100%', fig.cap="Fig.1: Omdat bij de per sequence quality scores de piek aan de rechterkant zitten kunnen we conluderen dat de reads een hoge gemiddelde Phred score hebben. De per base sequencing quality is van alle fastqc.html bastanden gemiddeld boven de 28 en dus binnen de verwachtte q-scores."}
# put fastqc results in this document
img1 <-  rasterGrob(as.raster(readPNG("Knipsel.PNG")))
img2 <-  rasterGrob(as.raster(readPNG("Knipsel2.PNG")))
grid.arrange(img1, img2, ncol=2, top=textGrob("Per base sequencing quality and per sequence quality of two fastqc .html outputs", gp=gpar(fontsize=10,font=8)))
```

### onderdeel 3 maak count table
```{r eval=FALSE}
# Create object with input dir for bam files
bam_dir <- "/home/daur2/rnaseq/rnaseq_onecut/bam/"

# Create object with output dir for count tables
counts_dir <- "/home/daur2/rnaseq/rnaseq_onecut/counts/"

# Create vector with names of bam files
bam_files <- list.files(bam_dir, pattern = c(SRR7866699, SRR7866700
, SRR7866705, SRR7866706), full.names = TRUE)

# Count the reads per gene using the in-built NCBI RefSeq annotations
read_counts <- featureCounts(
  
  files = bam_files,
  annot.inbuilt = "hg38",
  useMetaFeatures = TRUE,
  strandSpecific = TRUE,
  isPairedEnd = TRUE, 
  countReadPairs = TRUE, 
  nthreads = 10
)
```

```{r eval=FALSE, include=FALSE}
### onderdeel 0 controleer count table

# obtain list from directory
read_counts <- readRDS("/home/daur2/rnaseq/rnaseq_onecut/counts/read_counts.rds")

# Obtain the dataframe with the statistics
count_stats <- read_counts$stat

# Convert first column to rownames
rownames(count_stats) <- count_stats$Status
count_stats$Status <- NULL

# Calculate the percentage of uniquely counting fragments
count_stats_t <- count_stats %>% 
  t %>% 
  as_tibble() %>% 
  mutate(bamfile=colnames(count_stats)) %>%
  mutate(Total=colSums(count_stats)) %>%
  mutate(perc_assigned = Assigned/Total*100)

# Plot the percentages in a bar graph
count_stats_t %>% ggplot(aes(x = bamfile, y = perc_assigned)) +
  geom_col() +
  ggtitle("Proportion of counted fragments for each sample") +
  xlab("RNA-seq sample") +
  ylab("Percentage of counted fragments") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  coord_cartesian (ylim =c(0,100))
```

### onderdeel 4 maak DESeqDataSet
```{r}
# obtain list from directory
read_counts <- readRDS("/home/daur2/rnaseq/rnaseq_onecut/counts/read_counts.rds")

# obtain count matrix
count_matrix <- as.data.frame(read_counts$counts) %>% dplyr::select(SRR7866699.bam, SRR7866700.bam, SRR7866705.bam, SRR7866706.bam)

# Import the sample data
metadata <- read_csv("/home/daur2/rnaseq/rnaseq_onecut/onecut_sampledata.csv")
metadata <- metadata %>% filter(Run=="SRR7866699" | Run=="SRR7866700" | Run=="SRR7866705" | Run=="SRR7866706")

# Convert the metadata to dataframe object
metadata <- as.data.frame(metadata)

# Add rownames to the metadata dataframe
rownames(metadata) <- paste0(metadata$Run, ".bam")

# Check if column names of count table are the same as row names of metadata object
colnames(count_matrix) == rownames(metadata)

# Create a column specifying the test condition
metadata <- metadata %>% mutate(Cell_type = str_replace(Cell_type, "Skin derived fibroblast overexpressing Bclxl", "BCLXL"))
metadata <- metadata %>% mutate(Cell_type = str_replace(Cell_type, "2 days after induction of OC3 in skin derived fibroblasts", "OC3"))
metadata$Cell_type <- metadata$Cell_type %>% factor(levels = c("BCLXL", "OC3"))

# Create the DESeqDataSet object
dds <- DESeqDataSetFromMatrix(
  countData = count_matrix,
  colData = metadata, 
  design = ~ Cell_type
)
dds
# Normalise the DESeqDataSet object
dds_normalized <- rlog(dds)
```

### onderdeel 5a PC staafdiagram
```{r}
# Perform PCA on the airway study dataset
pca <- dds_normalized %>% assay() %>% t() %>% prcomp()

# Find the proportion of variance explained by each PC
pca_summary <- summary(pca)$importance

# Create data frame with metadata and the PCA values for plotting
pca_plotting <- cbind(metadata, pca$x)

# Obtain the percentages of variation covered by PC1 and PC2
PC1 <- round(pca_summary["Proportion of Variance", "PC1"]*100, digits = 1)
PC2 <- round(pca_summary["Proportion of Variance", "PC2"]*100, digits = 1)
PC3 <- round(pca_summary["Proportion of Variance", "PC3"]*100, digits = 1)
PC4 <- round(pca_summary["Proportion of Variance", "PC4"]*100, digits = 1)
PC_data <- data.frame(PC1, PC2, PC3, PC4)
PC_tidy <- PC_data %>% pivot_longer(cols = c("PC1", "PC2", "PC3", "PC4"),
                      names_to = "PC_number", values_to = "percentage")
PC_tidy %>%
  ggplot(aes(x=PC_number,y=percentage)) +
  geom_bar(stat="identity")+
  coord_cartesian(ylim=c(0,100))+
  theme_classic()+
  labs(title = "Variatiepercentage per principal component",
       x="PC number",
       y="Percentage") +
  theme(legend.position = "none", text = element_text(size=14))
```

### onderdeel 5b PCA
```{r}
# Plot PC1 vs PC2
ggplot(pca_plotting) + 
  geom_point(aes(x=PC1, y=PC2, color = Cell_type), size = 5) +
  ggtitle("PCA for airway study") +
  xlab(paste0("PC1 (", PC1, "%)")) +
  ylab(paste0("PC2 (", PC2, "%)")) +
  theme_bw()
```

Er is uit de PCA te concluderen dat PC1 ongeveer 93 procent van de data representeert en hier duidelijk twee groepen zijn en dat hierbij het celltype anders is, PC2 representeert ongeveer 4 procent van de data en  splitst de groepen niet op een heel duidelijke manier.

### onderdeel 6 DGE analyse
```{r}
# Perform the DGE analysis using DESeq2
onecut_dge <- DESeq(dds)

# Obtain the results for the DGE analysis
onecut_dge_results <- results(onecut_dge, alpha=0.01, lfcThreshold=1)

# Obtain the summary of the results
summary(onecut_dge_results)
```

### onderdeel 6a volcano plot
```{r}
# Create a dataframe for plotting without genes with padj = NA
onecut_dge_plotting <- data.frame(onecut_dge_results) %>% filter(!is.na(padj))

# Create column specifying if gene is significantly differentially expressed
onecut_dge_plotting <- onecut_dge_plotting %>% 
  mutate(signif = if_else(padj < 0.01 & abs(log2FoldChange) > 1, "Significant", "Not significant"))

# Create a volcano plot
onecut_dge_plotting %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), color = signif)) +
  geom_point() + 
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") + 
  theme_bw() +
  
# Change the legend text and the colours
scale_colour_manual(values = c("grey", "darkorange"), name = "Significance") +

# Indicate thresholds as lines
geom_hline(yintercept = -log10(0.01), linetype = "dashed") +
geom_vline(xintercept = 1, linetype = "dashed") +
geom_vline(xintercept = -1, linetype = "dashed") +
  
annotate("text", x = -6, y = 150, label = "644 down") +
annotate("text", x = -6, y = 137, label = "gereguleerd") +
annotate("text", x = 7.5, y = 150, label = "661 upgereguleerd")
```

### onderdeel 6b heatmap met gensymbool
```{r}
onecut_dge_filter <- as.data.frame(onecut_dge_results) %>% filter(onecut_dge_results$padj < 0.01)

# Create a dataframe for the 10 most significantly DE genes
top5_genes <- onecut_dge_filter[order(onecut_dge_filter$padj)[1:5],] %>% data.frame()
down5_genes <- onecut_dge_filter[order(onecut_dge_filter$padj, decreasing=TRUE)[1:5],] %>% data.frame()

# Create an extra column with the Entrez id (based on rownames)
top5_genes <- top5_genes %>% mutate(entrezid = rownames(top5_genes))
down5_genes <- down5_genes %>% mutate(entrezid = rownames(down5_genes))


# Obtain the gene symbols based on the Entrez identifiers
top5_genes$symbol <- mapIds(org.Hs.eg.db,
                             keys = top5_genes$entrezid,
                             column = "SYMBOL",
                             keytype = "ENTREZID",
                             multiVals = "first")

# Obtain the gene symbols based on the Entrez identifiers
down5_genes$symbol <- mapIds(org.Hs.eg.db,
                             keys = down5_genes$entrezid,
                             column = "SYMBOL",
                             keytype = "ENTREZID",
                             multiVals = "first")

# merge dataframes
topdown5_genes <- rbind(top5_genes, down5_genes)
topdown5_genes

# Obtain the count values for these genes
count_value <- assay(dds)[topdown5_genes$entrezid,]

# Replace the column names with conditions
colnames(count_value) <- colData(dds)$Cell_type

# Replace the row names with symbol
rownames(count_value) <- topdown5_genes$symbol

# Visualize the results in a heatmap
pheatmap(count_value, scale = "row", show_rownames = TRUE)
```

### onderdeel 7 functie maken
```{r}
en_uni_sym <- function(entrezid){
eus <- select(org.Hs.eg.db,
                                  keys = (entrezid),
                                  column = c("ENSEMBL", "UNIPROT", "SYMBOL"),
                                  keytype = "ENTREZID",
                                  multiVals = "list")
as_tibble(eus)
}
en_uni_sym(c("3175","9480","390874"))
```

### onderdeel 8 GO-term analyse
```{r}
# Create a list of upregulated genes
upregulated_genes <- onecut_dge_results %>% data.frame() %>% 
  filter(log2FoldChange > 1, padj < 0.01) %>% rownames()

# Create a list of all genes in the dataset
all_genes <- onecut_dge_results %>% data.frame() %>% rownames()

# Perform GO term enrichment analysis
test_object <- new("GOHyperGParams",
                   geneIds = upregulated_genes,
                   universeGeneIds = all_genes, 
                   annotation = "org.Hs.eg.db", 
                   ontology = "BP", 
                   pvalueCutoff = 1,
                   testDirection = "over")

goterm_analysis <- hyperGTest(test_object)
goterm_analysis

# Obtains dataframe with results of GO term analysis
goterm_analysis_results <- summary(goterm_analysis)

# Adjust the p values for multiple testing
goterm_analysis_results$padj <- p.adjust(goterm_analysis_results$Pvalue, method = "BH")

# Select only gene sets that are larger than 5 but smaller than 500 (to prevent taking along very small and very large gene sets).
goterm_analysis_results <- goterm_analysis_results %>% filter(Count > 5) %>% filter(Count < 500)

# Select the top 20 GO terms
goterm_analysis_top20 <- goterm_analysis_results[order(goterm_analysis_results$padj)[1:20],]

# Plot the p-values of the top 20 GO terms
goterm_analysis_top20$Term <- factor(goterm_analysis_top20$Term, 
                                     levels = goterm_analysis_top20$Term[
                                       order(goterm_analysis_top20$padj, decreasing = TRUE)])

goterm_analysis_top20 %>% ggplot(aes(x = Term, y = -log10(padj))) +
  geom_point() +
  coord_flip() +
  ylab(expression(-log[10](adjusted~italic(P)~value))) + 
  xlab("GO terms") +
  ggtitle("Top 20 enriched GO terms\nfor upregulated genes") +
  theme_bw()

###############################################

# Create a list of downregulated genes
downregulated_genes <- onecut_dge_results %>% data.frame() %>% 
  filter(log2FoldChange < -1, padj < 0.01) %>% rownames()

# Create a list of all genes in the dataset
all_genes <- onecut_dge_results %>% data.frame() %>% rownames()

# Perform GO term enrichment analysis
test_object <- new("GOHyperGParams",
                   geneIds = downregulated_genes,
                   universeGeneIds = all_genes, 
                   annotation = "org.Hs.eg.db", 
                   ontology = "BP", 
                   pvalueCutoff = 1,
                   testDirection = "over")

goterm_analysis <- hyperGTest(test_object)
goterm_analysis

# Obtains dataframe with results of GO term analysis
goterm_analysis_results <- summary(goterm_analysis)

# Adjust the p values for multiple testing
goterm_analysis_results$padj <- p.adjust(goterm_analysis_results$Pvalue, method = "BH")

# Select only gene sets that are larger than 5 but smaller than 500 (to prevent taking along very small and very large gene sets).
goterm_analysis_results <- goterm_analysis_results %>% filter(Count > 5) %>% filter(Count < 500)

# Select the top 20 GO terms
goterm_analysis_top20 <- goterm_analysis_results[order(goterm_analysis_results$padj)[1:20],]

# Plot the p-values of the top 20 GO terms
goterm_analysis_top20$Term <- factor(goterm_analysis_top20$Term, 
                                     levels = goterm_analysis_top20$Term[
                                       order(goterm_analysis_top20$padj, decreasing = TRUE)])

goterm_analysis_top20 %>% ggplot(aes(x = Term, y = -log10(padj))) +
  geom_point() +
  coord_flip() +
  ylab(expression(-log[10](adjusted~italic(P)~value))) + 
  xlab("GO terms") +
  ggtitle("Top 20 enriched GO terms\nfor downregulated genes") +
  theme_bw()
```